### ISSUE TITLE
Unspecified behavior from new-to-DXIL idiv instruction

### ISSUE THREAD

**Issue by jenatali** [microsoft/DirectXShaderCompiler#3150]
DXBC had a udiv instruction, with clearly-spec'd behavior for corner cases like divide by 0 (should produce maximum uint).

DXIL introduces an idiv instruction, but doesn't specify what should happen with:
* Positive value / 0
* Negative value / 0
* INT_MIN / -1, which would evaluate to INT_MAX+1, but that can't be represented

**Comment by llvm-beanz** [comment]
I assume division by zero is undefined in HLSL as it is in C/C++. If not, it probably should be (and effectively is now).

I think `INT_MIN / -1` overflows and is thus also undefined behavior.

LLVM IR and SPIR-V both specify that division by zero is undefined:
* LLVM Lang Ref - [`udiv`](https://llvm.org/docs/LangRef.html#udiv-instruction) & [`sdiv`](https://llvm.org/docs/LangRef.html#sdiv-instruction)
* [SPIR-V Spec](https://registry.khronos.org/SPIR-V/specs/1.0/SPIRV.pdf) See OpUDiv & OpSDiv on page 125.

My take here is that we should document this behavior, but not change anything.

@tex3d, @pow2clk any thoughts here?

**Comment by damyanp** [comment]
@jenatali - is the dxbc converter generating `idiv` instructions (we can't find one, so presumably this means `sdiv`?)  Or is the dxbc generating the dxil `udiv` operation (as opposed to the `udiv` LLVM IR instruction)?

**Comment by jenatali** [comment]
Ah apparently I had the wrong name, it's `sdiv`, not `idiv`. I'm not seeing this from the DXBC converter, but it's easy to hit: https://godbolt.org/z/bnTWGecr4

Now that you brought it up, I do actually see that the DXBC converter generates the `udiv` operation, as opposed to the `udiv` LLVM instruction, which means that those probably can/do have different semantics.

**Comment by llvm-beanz** [comment]
I've written a test to figure out what the compiler actually does for this over in my fancy WoL test framework:

https://github.com/llvm-beanz/HLSLTest/blob/main/test/Basic/idiv.yaml

What I've discovered is that through lack of testing we've basically inherited the undefined behavior here.

My test asserts that:
1/0 = -1
-1/0 = -1
INT_MAX / 0 = -1
INT_MIN / 0 = -1
0/0 = -1

1/-1 = -1
-1/-1 = 1
INT_MAX / -1 = INT_MIN + 1
INT_MIN / -1 = INT_MIN
0/-1 = 0

I've tested this on NVIDIA DirectX, NVIDIA Vulkan, Intel DirectX, Intel Vulkan and Apple. All tested platforms handle division by -1 the same.

Division by zero has different behavior on different drivers. The NVIDIA DirectX driver, and both NVIDIA and Intel Vulkan drivers return -1 for all division by zero. Apple returns -1 for all cases except 0/0 which it returns 0 for. The Intel DirectX driver returns INT_MAX when the numerator is >= 0 and INT_MIN when the numerator is < 0.

**Comment by llvm-beanz** [comment]
I guess one other thing worth noting is that LLVM constant evaluates division-by-zero to undef if both the numerator and denominator are constants. If just the denominator is constant it leaves the instruction unevaluated (which is odd to me):
https://godbolt.org/z/Ef6bqb9es

**Comment by damyanp** [comment]
Plan: add a note in DXIL.rst that sdiv divide by zero behavior is undefined.

**Comment by llvm-beanz** [comment]
In DXIL we have both the LLVM integer division instructions `udiv` and `sdiv`, and a UDiv DXIL operation (https://github.com/Microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#udiv).

The `UDIV` dxil operation is defined to match the DXBC `UDIV` instruction, while the `udiv` and `sdiv` instruction behavior is undefined. This is particularly meaningful because if you compile HLSL with DXC, integer division-by-zero is _undefined_. The full breath of divide by zero for HLSL division by zero is:

* FXC -> defined
* DXC (dxil) -> undefined
* DXC (spirv) -> undefined
* Clang -> undefined

The TL;DR is that division by zero is undefined.

**Comment by jenatali** [comment]
Ah, I think I missed that there's both a DXIL UDiv and LLVM div operators. That makes more sense.

**Comment by jholewinski** [comment]
> The TL;DR is that division by zero is undefined.

The DXIL ExecutionTest::BinaryUintOpTest#UDiv test verifies that implementations return -1 for div-by-zero: https://github.com/microsoft/DirectXShaderCompiler/blob/60a1c8568b405bbe9f47ab0431f627ef64988cc4/tools/clang/unittests/HLSLExec/ShaderOpArithTable.xml#L4518-L4572

Should we update this test? It looks like it may have been authored assuming DXC would produce a DXIL UDiv call instead of an llvm `udiv`/`urem`.

**Comment by llvm-beanz** [comment]
We had a discussion about this issue yesterday. At the time I wrote my comments above I was unaware that we had any conformance  testing in place that ensured drivers treated the `udiv` instruction in a way that matched DXBC's `udiv`. @jholewinski has helpfully found and pointed out the testing (Thank you!).

This is somewhat of a worst-case situation we're now in.

At some point we decided that the LLVM `udiv` instruction should be treated by drivers as if it is the DXBC `udiv` instruction, and we wrote some tests to enforce that behavior. There are a couple big problems with changing the semantic meaning of an LLVM instruction like this in DXIL. One is that we didn't actually update DXC's LLVM optimizations to respect the new meaning. For example code like this in Constant Folder assumes divide by zero is undef:

https://github.com/microsoft/DirectXShaderCompiler/blob/1866977946e62da49d0b69c9d39585b0740db90b/lib/IR/ConstantFold.cpp#L1020-L1025

Similarly LLVM's ValueTracking analysis also makes assumptions about `udiv`:

https://github.com/microsoft/DirectXShaderCompiler/blob/1866977946e62da49d0b69c9d39585b0740db90b/lib/Analysis/ValueTracking.cpp#L3072-L3078

The later case actually has a cascading impact because ValueTracking is used by other optimizations to determine reachability of control flow, and range of loops.

The other potentially bigger problem that changing the semantic meaning of an LLVM instruction has is on DXIL consumers (which are disproportionately newer versions of LLVM. Unsurprisingly new versions of LLVM assume `udiv` behaves like the LLVM instruction, so even if we fixed all the issues in DXC handling `udiv`, we would also need all DXIL consumers to do the same.

@jholewinski suggested three options based on our discussion yesterday:
1) Keep the existing behavior.
2) Update DXC to emit the DXIL UDiv instruction.
3) Match LLVM behavior for the instructions (undefined).

Option 1 is the lowest risk, but leaves us in a bad situation. Neither DXC nor modern LLVM are going to model `udiv` matching DXBC's behavior which we expect drivers to match. This could lead to a wack-a-mole bug problem. That said, DXC has been in use for a long time, and we've generally not encountered user-reported issues about this... so maybe we can hold our breath and hope it works out until SM 7 when we move to SPIRV and can have a bit of a clean break.

Option 2 is risky and is a whole lot of work to do correctly in DXC. This would require updating DXC's codegen layer to generate an IR intrinsic call instead of the `udiv` instruction. We would also need to update constant folding, value analysis and any other optimization and analysis passes that we need to respect the DXIL intrinsic otherwise we would have potentially significant behavior changes and performance regressions across existing shaders. Ultimately this is a lot of work to do in DXC (and Clang), and I personally am not sure this is worth it.

Option 3 is a lot less work, but poses significant risk to existing shaders and drivers. This option would basically involve us specifying `udiv` and `sdiv` divide-by-zero as undefined behavior (matching LLVM), removing/updating the testing, and hoping it doesn't break too many shipped games. We could gate this under a new shader model to limit the exposure to just new shaders, but that still has some drawbacks because a user recompiling their existing shader for a new shader model _might_ see a failure, or it might work. Since the current behavior would be a valid interpretation of undefined behavior the user would almost certainly struggle to identify bugs as a result of this change.

IMO, option 1 is our least-bad choice here unless we end up with a really strong motivation to pursue option 2.

I don't think option 2 is something we want to pursue because it is a whole ton of work and we really don't want to make this kind of investment in DXC unless we really need to. I'm also concerned about the long-term maintenance of this in modern LLVM as we move the compiler forward.

I think that option 3 is hostile to our users and should be discounted for that reason.

**Comment by jholewinski** [comment]
I suppose we could fold options 1 and 2 into a horrible hybrid: continue to emit `udiv`/`sdiv` in the codegen layer but add an explicit check for div-by-zero. Something like:

```llvm
begin:
  %0 = icmp ne i32 %den, 0
  br i1 %0, label %non-zero, %end

non-zero:
  %1 = udiv i32 %num, %den
  br label %end

end:
  %2 = phi i32 [ -1, %begin ], [ %1, %non-zero ]
```

As horrible as that is, IHVs may need to do something like it to properly implement the DXIL semantics. Ideally we could simplify it to something like:

```llvm
%0 = icmp eq i32 %den, 0
%1 = udiv i32 %num, %den
%2 = select i1 %0, i32 -1, i32 %1
```

But I don't think that's actually correct in modern LLVM, as the same issue could arise where LLVM infers %den cannot be zero due to the UB behavior.


My main concern with option 1 is the impact on using upstream LLVM tools to process DXIL. It would be nice to be able to run something like `opt -passes=...` on DXIL bitcode. But with option 1, we run the risk of upstream performing bad optimizations. IHVs would effectively be forced to adopt something like the above before any processing is done on the DXIL as LLVM IR. It's also unclear how this would work with modernization. Wouldn't Clang just fall into the same hole of incorrectly optimizing the divisions?

**Comment by jenatali** [comment]
Indeed WARP has the exact codegen you mention in its backend. Simply changing the code that DXC emits to include this additional control flow runs the risk of performance impact where the additional conditional logic might be inserted twice, unless it's done based on a shader model flag, which is effectively option 3.

**Comment by llvm-beanz** [comment]
> I suppose we could fold options 1 and 2 into a horrible hybrid: continue to emit `udiv`/`sdiv` in the codegen layer but add an explicit check for div-by-zero. Something like:

Separating HLSL from DXIL: if we wanted to make divide by zero defined in HLSL, this is more or less exactly what we should do. This would make it so that our initial IR generation is "correct", and would lower to SPIRV correctly (in Clang where we share codegen, but not necessarily DXC where we don't).

We could do this today in DXC and relax the DXIL testing for new shader models only. That would mean we would expect `udiv` to behave matching DXBC for SM 6.0->6.9, but maybe in 6.10 we say it is undefined, but since the compiler would generate handling for the UB the language-level behavior would remain defined.

**Comment by damyanp** [comment]
@llvm-beanz & I discussed this briefly - we're going to continue with option 1 ("keep the existing behavior") for the foreseeable future unless a compelling reason to change direction arises, and ensure that SM 7.0 doesn't end up in this situation.

**Comment by HansKristian-Work** [comment]
sdiv/srem behavior for div by 0 diverges on impls:

- NV/WARP: 0xfffffffff
- AMD: INT_MAX or INT_MIN depending on numerator sign.
